core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -165,6 +192,50 @@ LocalExchangeSource getSource(int partitionIndex)
         return sources.get(partitionIndex);
     }
 
+    private static PartitionFunction createPartitionFunction(
+            PartitioningProviderManager partitioningProviderManager,
+            Session session,
+            BlockTypeOperators blockTypeOperators,
+            PartitioningHandle partitioning,
+            int partitionCount,
+            List<Type> partitionChannelTypes,
+            boolean isHashPrecomputed)
+    {
+        if (partitioning.getConnectorHandle() instanceof SystemPartitioningHandle) {
+            HashGenerator hashGenerator;
+            if (isHashPrecomputed) {
+                hashGenerator = new PrecomputedHashGenerator(0);
+            }
+            else {
+                hashGenerator = new InterpretedHashGenerator(partitionChannelTypes, IntStream.range(0, partitionChannelTypes.size()).toArray(), blockTypeOperators);
+            }
+            return new LocalPartitionGenerator(hashGenerator, partitionCount);
+        }
+
+        ConnectorNodePartitioningProvider partitioningProvider = partitioningProviderManager.getPartitioningProvider(partitioning.getConnectorId().get());
+        ConnectorBucketNodeMap connectorBucketNodeMap = partitioningProvider.getBucketNodeMap(
+                partitioning.getTransactionHandle().orElse(null),
+                session.toConnectorSession(),
+                partitioning.getConnectorHandle());
+        checkArgument(connectorBucketNodeMap != null, "No partition map %s", partitioning);
+
+        int bucketCount = connectorBucketNodeMap.getBucketCount();
+        int[] bucketToPartition = new int[bucketCount];
+        for (int bucket = 0; bucket < bucketCount; bucket++) {
+            bucketToPartition[bucket] = bucket % partitionCount;
+        }
+
+        BucketFunction bucketFunction = partitioningProvider.getBucketFunction(
+                partitioning.getTransactionHandle().orElse(null),
+                session.toConnectorSession(),
+                partitioning.getConnectorHandle(),
+                partitionChannelTypes,
+                bucketCount);
+
+        checkArgument(bucketFunction != null, "No bucket function for partitioning: %s", partitioning);
```

Comment:
same thing as https://github.com/trinodb/trino/pull/6924/commits/33c03c054095eecf1c10f823bf7d6fd7d1da9599#r577762564. We should not need to validate here.